static/app/views/dashboards/manage/index.tsx

```diff
@@ -71,17 +84,85 @@ function ManageDashboards() {
         query: {
           ...pick(location.query, ['cursor', 'query']),
           sort: getActiveSort().value,
-          per_page: '9',
+          per_page: rows * columns,
         },
       },
     ],
     {staleTime: 0}
   );
 
-  const [showTemplates, setShowTemplatesLocal] = useLocalStorageState(
-    SHOW_TEMPLATES_KEY,
-    shouldShowTemplates()
-  );
+  const dashboardsPageLinks = getResponseHeader?.('Link') ?? '';
+
+  useEffect(() => {
+    const dashboardGridObserver = new ResizeObserver(
+      debounce(entries => {
+        entries.forEach(entry => {
+          const currentWidth = entry.contentRect.width;
+
+          setGridWidth(prevWidth => {
+            if (currentWidth !== prevWidth && prevWidth !== 0) {
+              setResizing(true);
+              return currentWidth;
+            }
+            return prevWidth;
+          });
+
+          setRowsAndColumns(currentWidth);
+
+          const paginationObject = parseLinkHeader(dashboardsPageLinks);
+          if (
+            dashboards?.length &&
+            paginationObject.next.results &&
+            rows * columns > dashboards.length
+          ) {
+            refetchDashboards();
+          }
+        });
+      }, 10)
+    );
+
+    if (dashboardGridRef.current) {
+      dashboardGridObserver.observe(dashboardGridRef.current);
+    }
+
+    return () => {
+      if (dashboardGridRef.current) {
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+        dashboardGridObserver.unobserve(dashboardGridRef.current);
+      }
+      setResizing(false);
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  // no longer resizing if loading has stopped
+  useEffect(() => {
+    if (resizing && !isLoading) {
+      setResizing(false);
+    }
+  }, [isLoading, resizing]);
+
+  function setRowsAndColumns(containerWidth: number) {
+    const widgetWidth = 300;
+    const padding = 16;
+    const defaultNumRows = 3;
+    const defaultNumWidgets = 8;
```

Comment:
These constants should be declared somewhere higher up! Since this is the source of truth for how the grid behaves, all code should refer to them.

e.g., since the padding is defined here, the CSS in `DashboardsList` should use it, too. Same with the widget width, etc. The default number of rows and columns should be shared with the default state of the `rows` and `columns` state